litestar/middleware/cors.py

```diff
@@ -68,15 +78,55 @@ async def wrapped_send(message: Message) -> None:
                     headers["Access-Control-Allow-Origin"] = origin
                     headers["Vary"] = "Origin"
 
-                # We don't want to overwrite this for preflight requests.
-                allow_headers = headers.get("Access-Control-Allow-Headers")
-                if not allow_headers and self.config.allow_headers:
-                    headers["Access-Control-Allow-Headers"] = ", ".join(sorted(set(self.config.allow_headers)))
+                headers["Access-Control-Allow-Headers"] = ", ".join(sorted(set(self.config.allow_headers)))
 
-                allow_methods = headers.get("Access-Control-Allow-Methods")
-                if not allow_methods and self.config.allow_methods:
-                    headers["Access-Control-Allow-Methods"] = ", ".join(sorted(set(self.config.allow_methods)))
+                headers["Access-Control-Allow-Methods"] = ", ".join(sorted(set(self.config.allow_methods)))
 
             await send(message)
 
         return wrapped_send
+
+    def create_preflight_response(self, origin: str, request_headers: Headers) -> Response[str | None]:
```

Comment:
is this meant to be private?

Reply:
It's an interesting question, because while this class is public interface, there isn't really any way for users to customize the way it works. So IMO this whole class should be private.

Prior to this PR, there needed to be a `CORSConfig` passed to the application object in order for pre-flight responses to be generated:

https://github.com/litestar-org/litestar/blob/a0d11092dae300af4025d619d8ac818790dede74/litestar/routes/http.py#L258-L262

There is no way to only supply the `CORSMiddleware` to a subset of routes via a `Router.middleware` or similar, because of the fact that the pre-flight response generation is conditional upon the application having access to the `CORSConfig` object. In other words, if a user didn't supply the `CORSConfig` object to the application, but did supply `DefineMiddleware(CORSMiddleware, config=CORSConfig(...))` to a router, then the pre-flight responses for that router wouldn't have worked anyway because there's no `CORSConfig` on the app. And if the `CORSConfig` is added to the app, then there is no point adding the middleware to a sub-router, because we'd then apply it to the whole application anyway.

Also, the `CORSMiddleware` class is hard coded as a wrapper around the `ASGIRouter`:

https://github.com/litestar-org/litestar/blob/a0d11092dae300af4025d619d8ac818790dede74/litestar/app.py#L842-L844

There is no way for a user to sub-class the middleware to alter its behavior, as there is no way to pass that sub-classed middleware to us.

So IMO, `CORSMiddleware` is implementation detail - the whole thing should be undocumented and we should probably deprecate it from public use.

Reply:
@litestar-org/maintainers, can I deprecate `CORSMiddleware` from our public interface based on :arrow_up:?